### ISSUE TITLE
PR #3 va #4 da test DAT - con 3 viec nho truoc khi sang TASK 09

### ISSUE THREAD

**Issue by Phuthanh123456** [Phuthanh123456/bku-project#5]
Chào Quân và Bảo, anh đã pull cả PR #3 và PR #4 về chạy thử kỹ rồi. Nói trước cho hai đứa yên tâm: **cả hai PR đều đạt hết tiêu chí, không có lỗi nào chặn, anh sẽ merge.**

Anh còn thử ghép code của hai đứa lại với nhau thì merge sạch, không xung đột, ráp được một lớp encoder hoàn chỉnh chạy thông, gradient chảy qua cả hai phần, không NaN.

---

## Kết quả test

**Quân — TASK 05 + 06**

| Kiểm tra | Yêu cầu | Đạt được |
|---|---|---|
| Đối chiếu `F.scaled_dot_product_attention` | < 1e-5 | **5,6e-17** |
| RoPE giữ độ dài vector | < 1e-5 | 8,5e-08 |
| RoPE vị trí 0 là phép đồng nhất | = 0 | **0** |
| RoPE chỉ phụ thuộc hiệu (m−n) | < 1e-5 | 2,4e-06 |
| Cross-attention không dính RoPE (bài 11) | giống hệt | **lệch 0** |
| Trọng số attention hợp lệ (bài 3) | tổng = 1, ≥ 0, = 0 ở mask | đạt cả ba |
| Causal mask không cho nhìn về sau (bài 1) | không đổi | **lệch 0** |
| Che toàn bộ một hàng ở fp16 | không NaN | không NaN, không Inf |
| Xoay cả chuỗi == xoay từng bước (KV cache) | bằng nhau | **lệch 0** |

**Bảo — TASK 07 + 08**

| Kiểm tra | Yêu cầu | Đạt được |
|---|---|---|
| Đối chiếu `torch.nn.RMSNorm` | < 1e-5 | **2,9e-08** |
| RMSNorm trung bình bình phương = 1 (bài 8) | < 1e-5 | 2,9e-07 |
| Cộng hằng số vào thì kết quả đổi (khác LayerNorm) | phải đổi | đổi 2,43 |
| SwiGLU vs công thức NumPy (bài 9) | khớp | **2,8e-17** |
| Số tham số SwiGLU vs ReLU | chênh < 1% | **0,78%** |
| SwiGLU = 3 × 512 × 688 | 1.056.768 | đúng |
| Đầu vào 300.0 ở fp16 (bình phương tràn) | không NaN | không NaN, đầu ra vẫn khác 0 |
| Pre-Norm với sublayer = 0 → đường residual sạch | ra == vào | **lệch 0** |

Bảo làm thêm `ResidualConnection` và `build_final_norm` ngoài yêu cầu — hai cái này coi như đã làm sẵn **bài kiểm tra số 10** cho TASK 10 của My luôn. Ghi nhận.

Anh cũng thử dựng mô hình từ cả 5 file cấu hình ablation thật (`base`, `a1_layernorm`, `a4_sincos`, `a5_relu`, `a6_post_norm`) thì đều ra đúng tổ hợp mong đợi. Nghĩa là phần ablation của anh ở TASK 17 chạy được.

---

### 1. TASK 09 phải gọi đúng tên hàm Bảo đã viết

Bảo đổi tên các hàm factory so với stub gốc. Em dùng đúng danh sách này, đừng viết lại:

```python
from nmt.model.normalization import build_norm, ResidualConnection, build_final_norm
from nmt.model.layers        import build_ffn, EncoderLayer, DecoderLayer
from nmt.model.attention     import MultiHeadAttention
from nmt.model.positional    import RoPE, MaHoaViTriSinCos
```

Tên cũ trong stub (`tao_chuan_hoa`, `tao_ffn`, `LopEncoder`, `LopDecoder`) **không còn nữa**. Anh đã ghi sẵn danh sách này vào đầu file `src/nmt/model/transformer.py` rồi, em mở ra là thấy.

Hai chỗ dễ sai:
- `ResidualConnection` đã xử lý sẵn **cả Pre-Norm lẫn Post-Norm**, nên trong `EncoderLayer`/`DecoderLayer` **không cần viết nhánh `if` nào** cho `vi_tri_chuan_hoa`.
- Mỗi khối con cần **một `ResidualConnection` riêng**. Dùng chung một cái là hai khối xài chung trọng số chuẩn hóa — chương trình vẫn chạy, mô hình vẫn học, nhưng kiến trúc đã sai mà không có gì báo lỗi.

### 2. Còn thiếu hàm chọn mã hóa vị trí theo cấu hình

Tiêu chí "Xong khi" của TASK 06 là *"đổi cấu hình là chuyển được giữa 2 phương án mà không sửa code"*. Hiện `RoPE` và `MaHoaViTriSinCos` đều viết xong và đều đúng, nhưng **chưa có hàm nào đọc `cfg.mo_hinh.ma_hoa_vi_tri` để chọn**, nên chưa chứng minh được tiêu chí đó.

Em thêm giúp anh một hàm kiểu:

```python
def build_positional(cfg):
    """Đọc cfg.mo_hinh.ma_hoa_vi_tri -> RoPE hoặc MaHoaViTriSinCos."""
```

Đặt trong `positional.py` cho đồng bộ với `build_norm` và `build_ffn` của Bảo. Thiếu hàm này thì **ablation A4 (RoPE vs sin-cos) của anh ở TASK 17 không chạy được**, nên làm luôn trong TASK 09 nha.

---

### 3. Sửa 2 dòng chú thích trong PR #4

Cái này nhỏ xíu, em sửa rồi push lại lên nhánh cũ là được:

**a)** Đầu file `src/nmt/model/normalization.py` đang ghi `Người làm: Quân` — theo bảng phân công cuối thì TASK 07 là của em, sửa thành `Người làm: Bảo`.

**b)** Docstring của `build_final_norm` viết:

> *"Với Post-Norm thì lớp cuối cùng của chồng đã tự chuẩn hóa rồi nên trả về `nn.Identity`"*

Nhưng code thực tế trả `nn.Identity` khi `use_final_norm=False`, bất kể `pre` hay `post`. Anh kiểm hai đường cấu hình thật thì **đều chạy đúng** (`base.yaml` → RMSNorm, `a6_post_norm` → Identity), nên chỉ cần sửa lại câu docstring cho khớp code thôi, không phải sửa logic.

---

## Ghi chú chung cho cả hai

**Đừng gọi `model.half()`** trong Training Loop hay bất kỳ đâu. Dự án dùng `torch.autocast` + `GradScaler`. Nếu gọi `.half()` thì bảng góc quay `cos_cache`/`inv_freq` của RoPE bị ép xuống fp16 và mất độ phân giải vị trí — đúng cái bẫy fp16 số 2 mà cả hai đã cẩn thận tránh được ở tầng code.

**Anh đã sửa sẵn tên người phụ trách** trong các file stub còn lại (`transformer.py`, `checkpoint.py`, `hub_sync.py`, `loss.py`, `scheduler.py`, `trainer.py`) theo bảng phân công cuối, đã push lên `main`. Hai đứa nhớ `git pull` trước khi làm tiếp.

**Nhắc lại bảng phân công cuối cùng:**

| | Task |
|---|---|
| **Quân** | 05 ✅ · 06 ✅ · **09** · **11** · **13** |
| **Bảo** | 07 ✅ · 08 ✅ · **12** · **14** · **15** |

Có gì không rõ cứ nhắn anh nha.
